base/compiler/typeinfer.jl

```diff
@@ -66,6 +66,36 @@ Timing(mi_info, start_time) = Timing(mi_info, start_time, start_time, UInt64(0),
 
 _time_ns() = ccall(:jl_hrtime, UInt64, ())  # Re-implemented here because Base not yet available.
 
+# A buffer of completed type inference timing trees.
+# GUARDED BY: jl_typeinf_profiling_lock
+const _finished_timings = Timing[]
+
+"""
+    Core.Compiler.Timings.clear_and_fetch_timings()
+
+Empty out the previously recorded type inference timings (`Core.Compiler._timings`).
+"""
+function clear_and_fetch_timings()
+    ccall(:jl_typeinf_profiling_lock_begin, Cvoid, ())
+    try
+        timings_copy = copy(_finished_timings)
+        empty!(_finished_timings)
+        return timings_copy
+    finally
+        ccall(:jl_typeinf_profiling_lock_end, Cvoid, ())
+    end
+    return nothing
+end
+
+function finish_timing_profile(timing::Timing)
+    ccall(:jl_typeinf_profiling_lock_begin, Cvoid, ())
+    try
+        push!(_finished_timings, timing)
+    finally
+        ccall(:jl_typeinf_profiling_lock_end, Cvoid, ())
+    end
+end
```

Comment:
These functions should be implemented in C, and in general we should not take C locks from Julia-level type inference.

Reply:
@pchintalapudi: See this discussion on slack where @vtjnash suggested we use a new C lock for this:
https://julialang.slack.com/archives/C688QKS7Q/p1668711395201969?thread_ts=1668710990.816479&cid=C688QKS7Q

Why would it be preferable to implement in C? I don't see any benefit to that.

Reply:
The reasoning is actually that there's a deadlock where thread A holds `typeinf_profiling_lock` and thread B holds the codegen lock, but thread A decides it needs to codegen `push!`, which requires holding the codegen lock, and thread B decides it needs to type infer something else, calling typeinf, and then blocks waiting for `typeinf_profiling_lock`.

Alternatively, we can look at it from the locking documentation perspective: https://github.com/JuliaLang/julia/blob/master/doc/src/devdocs/locks.md. `typeinf_profililng_lock` should be a level 2 lock since it guards a push that may GC, but julia code may not be called while holding anything under a level 6 lock. The only way to satisfy this criterion is to implement the method in C, which also avoids the deadlock since it's guaranteed that C will not trigger type inference or codegen.

Reply:
Yeah I wouldn't recommend trying that mostly because this is type inference and not user code, so the deadlock could happen anytime. If the lock was outside type inference, then it wouldn't matter and it could be written in Julia.

Reply:
yep, makes sense. Thanks a bunch! I've moved it to C (C++) in the latest commit. I'll also add this lock to that doc. 👍

Reply:
👍 k, added to the list as well. Thank you for the quick and careful review, @pchintalapudi! 👍 ❤️